core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@junrao Evaluation results show that the bulk of the time spent in processing ControlledShutdownRequest as of current Apache Kafka implementation is not in the synchronize writes to zookeeper. The optimization made in this patch can reduce the time of restarting a single broker by 10X (from 2-3 min to 15 seconds) in a real cluster of 4 brokers and 10K partitions with PROD traffic.

More specifically, there are two optimizations made in this patch. The first one will reduce number of LeaderAndIsrRequest sent by controller from O(partitionNum) to O(brokerNum) in order to handle one ControlledShutdownRequest. This optimization is proven to significantly reduce rolling bounce time (from 2-3 min to 15 seconds in a real cluster with traffic) and we have been using it in PROD for a few months. The second optimization reduces the number of MetadataUpdateRequest from O(partitionNum) to O(brokerNum) and some internal loop from O(partitionNum*brokerNum) to O(partitionNum + brokerNum). This is shown to reduce processing time of ControlledShutdownRequest by another 37% (from 3.8 sec to 2.4 in a cluster of 5K partitions without traffic).